pandas/core/groupby/groupby.py

```diff
@@ -3290,6 +3304,102 @@ def max(
                 npfunc=np.max,
             )
 
+    def _idxmin_idxmax(
+        self,
+        how: Literal["idxmax", "idxmin"],
+        numeric_only: bool = False,
+        skipna: bool = True,
+        ignore_unobserved: bool = False,
+    ) -> NDFrameT:
+        """Compute idxmax/idxmin.
+
+        Parameters
+        ----------
+        how: {"idxmin", "idxmax"}
+            Whether to compute idxmin or idxmax.
+        numeric_only : bool, default False
+            Include only float, int, boolean columns.
+        skipna : bool, default True
+            Exclude NA/null values. If an entire row/column is NA, the result
+            will be NA.
+        ignore_unobserved : bool, default False
+            When True and an unobserved group is encountered, do not raise. This used
+            for transform where unobserved groups do not play an impact on the result.
+
+        Returns
+        -------
+        Series or DataFrame
+            idxmax or idxmin for the groupby operation.
+        """
+        # seen is passed to the Cython code and mutated
+        seen = np.zeros(self.ngroups, dtype="bool")
+
+        def post_process(res):
+            has_na_value = (res._values == -1).any(axis=None)
+            has_unobserved = not seen.all()
+            raise_err: bool | np.bool_ = not ignore_unobserved and has_unobserved
+            if (
+                not self.observed
+                and not ignore_unobserved
+                and not raise_err
+                and (len(self.grouper.groupings) > 1 or res.empty)
+            ):
+                # When there are multiple groupings or the result is empty, the
+                # categories are not included in the upfront computation so we
+                # compare the final result length with the current length
+                result_len = np.prod(
+                    [len(ping.group_index) for ping in self.grouper.groupings]
+                )
+                raise_err = len(res) < result_len
+            if raise_err:
+                raise ValueError(
+                    f"Can't get {how} of an empty group due to unobserved categories. "
+                    "Specify observed=True in groupby instead."
+                )
+            elif not skipna and has_na_value:
+                warnings.warn(
+                    f"The behavior of {type(self).__name__}.{how} with all-NA "
+                    "values, or any-NA and skipna=False, is deprecated. In a future "
+                    "version this will raise ValueError",
+                    FutureWarning,
+                    stacklevel=find_stack_level(),
+                )
+
+            index = self.obj._get_axis(self.axis)
+            if res.size == 0:
+                result = res.astype(index.dtype)
+            else:
+                if isinstance(index, MultiIndex):
```

Comment:
Is this always safe to do? We won't ever lose data going from multi to single index?

Reply:
This converts the MulitIndex to tuples, which we need because they're going to wind up as column values.

I may have misunderstood your question.

Reply:
Probably my lack of understanding of groupby, but I was wondering if there was any a case where flattening the result of `self.obj._get_axix(self.axis)` produced tuples that shouldn't be included in the result set. As an example, if we had a MultIndex in the frame and grouped using the `levels` argument to groupby, that would still be safe here right?

Reply:
Yes - that's safe. self.obj is always the original object used to create the groupby, no modifications regardless of the groupby operation.

idxmin/idxmax return the indices from the origin object (`self.obj`).